src/Kernel/CacheInvalidatingContainer.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Core\Kernel;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+final class CacheInvalidatingContainer implements ContainerInterface {
+    public function __construct(
+        private readonly ContainerInterface $wrapped
+    )
+    {}
+
+    public function set(string $id, ?object $service): void
+    {
+        $this->wrapped->set($id, $service);
+    }
+
+    public function get(string $id, int $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE): ?object
+    {
+        try {
+            return $this->wrapped->get($id, $invalidBehavior);
+        } catch (\Throwable $throwable) {
+            // clear compiled container cache, to trigger re-discovery
+            RectorKernel::clearCache();
+
+            throw new \RuntimeException(
+                'Container cache is outdated and was cleared. please re-run the command.',
```

Comment:
lets throw with a more explicit error message, so people know about the "auto-healing" capabilities